### DISCUSSION TITLE
Phone app

### DISCUSSION THREAD

**Discussion by agdamsbo** [music-assistant/support#1834] [Feature requests and ideas]
Hi. So I have been following along for quite some time. The other day I stumpled upon the [android-squeezer](https://github.com/kaaholst/android-squeezer) Android app. This has opened some new very nice features for me. The android-squeezer app is intended as a LMS server remote control, but as MA is acting as an LMS server (as far as I understand), it also allows me to remote control playback on my different speakers and groups both from the app, but also outside the app, as it runs in the background. It is really good. It does not, however allow me to browse my library, which I would really like. I also have the HA-app, which of course allows for managing playback as well, but not as easily. Would it be possible to allow for browsing the library in an app like this? Or fork or modify the android-squeezer (apache 2.0 license) to make it compatible with MA?
Thank you for your great work! After b83, the experience is really great and quick playback actions. Thanks!

**Comment by OzGav** [comment]
There is currently a desktop app, mobile apps will be somewhere down the road.  Note that the current frontend does adapt to mobile screen sizes. What is the frontend not doing that this other app is doing?

**Reply by teranex** [reply]
> ...What is the frontend not doing that this other app is doing?

Having an Android app would be nice as it would allow integration with the notification + volume system, which makes it easier to quickly pause/resume playback on the target device and change the volume of that device.

**Comment by NMBRPL8** [comment]
Yes please!! I've also been hunting for some way to get this as a standalone android app, or at least as a dashboard component rather than its own tab on the sidebar on ha. Iframe is not ideal...  Having control via the android device physical buttons, and playback on the device also opens up Bluetooth playback in the car, on the go, just opens up the options significantly.

**Comment by pjoester** [comment]
Music Assistant seems to be a very promising project. One of the major things that prevents me switching from LMS to Music Assistant is that there is no way to use my physical volume buttons of my android phone. I use "Squeeze Ctrl" and I can adjust the volume while the phone screen is off. With Squeeze Ctrl Android also gets details about the media playback and one can see it on the standby/locked screen.

I dont' know if its possible to integrate such features directly into the Home Assistant App or if a standalone app is necessary. It would be great if one of these options could be possible in the future. Thanks for the awesome work!

**Comment by Vendo233** [comment]
I would be interested in standalone android apl as well

**Comment by jpbaril** [comment]
@agdamsbo How did you get to have this work (for at least playback control)?

The app supposedly found by itself my MA instance at http://homeassistantIP:9000 but then nothing is shown.

Thanks

**Reply by agdamsbo** [reply]
I think the underlying use of Logitech Media Server changed some time ago. I am not able to use the squeezer app anymore. I recently bought an IKEA TRETAKT plug with a remote. When paired with HA yhe remote works well for basic music control and the plug is great as a ZigBee router to extend coverage.

**Reply by jpbaril** [reply]
Okay, I understand why nothing was shown. What I found is that this Android-Squeezer remote app does not control the server itself, rather it controls the individual Squeeze clients currently connected to that server..
So in my case it is useless as I don't intend to have Squeeze clients.

**Comment by tvdruenen** [comment]
If you use Home Assistant as well next to Music Assistant, it works very well to add Music Assistant as a [Home Assistant Dashboard](https://www.home-assistant.io/dashboards/dashboards/#creating-a-new-dashboard).

This way you can access your Music Assistant library and controls directly from the Home Assistant UI and Companion apps. The UI scales properly on mobile phones.

Combining it with HACS plugins makes the controls even more easily accessible in your favourite existing dashboards:

- https://github.com/piitaya/lovelace-mushroom/blob/main/docs/cards/media-player.md, or
- https://github.com/kalkih/mini-media-player

**Reply by formatBCE** [reply]
It scales, but inconvenient as heck. Okay for web, but on mobile there's much better layouts possible.

**Comment by formatBCE** [comment]
So did anyone start doing it?
If not, I might start. Just don't want to do a monkey job, if someone is already about to publish it.

I could start it on KMP, that would allow both Android and iOS.

**Reply by OzGav** [reply]
That would be great, You should start a thread in the devs support first to get some advice from Marcel https://discord.com/channels/753947050995089438/1260489193763639338

**Reply by formatBCE** [reply]
> That would be great, You should start a thread in the devs support first to get some advice from Marcel https://discord.com/channels/753947050995089438/1260489193763639338

Oh yeah will ask him now.

**Reply by PablexXXXX** [reply]
I'd be happy to test and report everything.

**Reply by formatBCE** [reply]
There's app in dev right now. But it's broken with 2.7. working to get it back on rails.

**Comment by saarea** [comment]
I'm really new to both home assistant and music assistant but this project has blown my mind. It's truly amazing.  

Just thought I'd throw my 2 cents in here... I have no issue using music assistant but getting the rest of the family or guests to use it is a tough sell without an app. Getting them to log in via home assistant is scary for them and they don't really want to use it in a browser. If you had a dedicated app (even if it just wraps the web interface) it would be much easier to spread adoption. 

Thank you all and keep up the great work!

**Reply by formatBCE** [reply]
Well I just saved MA web page to shortcut, it looks like application on Android..:)

**Reply by saarea** [reply]
It looks like it... but it isn't, is it?  I can't wait to explain to my wife and parents how "it's just _like_ an app".  "They look _just like_ real diamonds, my love"

**Reply by formatBCE** [reply]
> It looks like it... but it isn't, is it?  I can't wait to explain to my wife and parents how "it's just _like_ an app".  "They look _just like_ real diamonds, my love"

Mobile app is in work. It's working, but there's no APK ready to install yet. You can checkout the repo and install it, if you have experience with Android Studio. :)

**Reply by prestonsmartuk** [reply]
> Well I just saved MA web page to shortcut, it looks like application on Android..:)

How are you doing this on Android? I'm trying to do the MA page as a shortcut, but every time I open the shortcut it opens up my default Home Assistant dashboard instead of Music Assistant

**Reply by formatBCE** [reply]
> > Well I just saved MA web page to shortcut, it looks like application on Android..:)
> 
> How are you doing this on Android? I'm trying to do the MA page as a shortcut, but every time I open the shortcut it opens up my default Home Assistant dashboard instead of Music Assistant

Oh, sorry, i have it in Docker, not as add-on. But i think you can use it via `http://HA_IP:8095`

**Comment by indykoning** [comment]
To add to this I would love the ability to use music assistant to play and control playback in Android auto. 
So I could drop the native Spotify etc. app

**Reply by formatBCE** [reply]
That would be super-nice!

I'm working on mobile app ow, but it's mostly players management app than player itself. I will think about player functionality - but probably it should be separate app (that brings Android device as a player to MA).

**Reply by jpbaril** [reply]
@formatBCE I don't know how Android APIs work but even if your app would only be to control external players, it would really be important to me that the playback be controlled from that app, and so that the playback controls of those players appear in typical Android locations, that is in the notification bar and lock-screen widget.

**Reply by formatBCE** [reply]
> @formatBCE I don't know how Android APIs work but even if your app would only be to control external players, it would really be important to me that the playback be controlled from that app, and so that the playback controls of those players appear in typical Android locations, that is in the notification bar and lock-screen widget.

That would be really nice, and it's probably doable. But there's bunch of unclear and restricted points.

E.g. I have 9 players in my home so far. Should I have media notification for every of them?

Or: what to do when I'm leaving home WiFi and MA isn't accessible anymore?

**Reply by jpbaril** [reply]
> E.g. I have 9 players in my home so far. Should I have media notification for every of them?

If I look at the Synology DS Audio mobile app, even if multiple players are currently playing, from the app you can only select to control one player/group at a time, and so that is the one that appears in the notifications and lock-screen.
That is the same in MA, at the bottom in the playback bar you only see one player at once.

BTW, it looks like Android is tricked by Synology DS Audio as to think that the playback is 'local', so in the widget it shows the music output to be "This phone".

> Or: what to do when I'm leaving home WiFi and MA isn't accessible anymore?

I suppose the controls would simply disappear.

**Reply by formatBCE** [reply]
> I suppose the controls would simply disappear.

And then re-appear on connection?

About one player visible at the time: it's the most inconvenient part of MA interface for me, and I'm changing that in app.

Need to think...

P.S. regards 'this phone' trick - I believe it's unintentional. They just create media session wrapper to display that notification, and since it's created locally, it belongs to phone...

**Reply by jpbaril** [reply]
> About one player visible at the time: it's the most inconvenient part of MA interface for me, and I'm changing that in app.

In my case I never control two completely different sets of players at the same time and even if I would I'm not sure why I would absolutely want to  see the two side by side in permanence. I mean, I can't be in two rooms and listen to two music streams at the same time... What would more likely happen is to control one group with multiple players and then enable/disable some individual players of that group.

For me the most important and frequent actions I perform are play/pause, next/previous, volume up/down. So I want these at hand. For the rest, I'm fine with opening the app and then having to do a few clicks to achieve my desired action.

**Reply by formatBCE** [reply]
> In my case I never control two completely different sets of players at the same time and even if I would I'm not sure why I would absolutely want to  see the two side by side in permanence. I mean, I can't be in two rooms and listen to two music streams at the same time... What would more likely happen is to control one group with multiple players and then enable/disable some individual players of that group.
> 
> For me the most important and frequent actions I perform are play/pause, next/previous, volume up/down. So I want these at hand. For the rest, I'm fine with opening the app and then having to do a few clicks to achieve my desired action.

Well all of this is achievable already. I have music player widgets from HA on my home screen, and I can use them for all the things you're describing.

**Reply by jpbaril** [reply]
I was referring to lock-screen (widget).

**Reply by formatBCE** [reply]
> I was referring to lock-screen (widget).

I understand. But making whole app only for media session notifications seems a bit too much. :)

Anyways. I'm making app now, but it's multiplatform. I'm Android dev, and I'll try to come up with comprehensive system for media notifications. But still - first app should be doing most stuff that you expect from it to do, like managing queue, for example.

**Reply by formatBCE** [reply]
I actually implemented all the things from this thread, except Android Auto so far. Case is, that connecting to MA needs VPN now, and Android Auto doesn't work while you're on VPN...

**Reply by jpbaril** [reply]
I just watched the YouTube vidéo demo you did a few weeks ago, great!
Is there an .apk somewhere?

**Reply by formatBCE** [reply]
Sorry, we still didn't manage to build artifacts, as we're deciding where to publish it. There's no signing certificates and stuff.
You may pull the repo and run it in Android Studio though.

**Reply by loryanstrant** [reply]
What are the artifacts you're referring to (just for my curiosity)?
I'm not a dev so am not familiar with the app development process.

**Reply by formatBCE** [reply]
> What are the artifacts you're referring to (just for my curiosity)?
> I'm not a dev so am not familiar with the app development process. 

I meant APK file in case of Android. 
Music Assistant adopted the app, and at some point the app will be published to Play store - but for now there's no other way to try it than build it from source...

**Reply by cociweb** [reply]
> I actually implemented all the things from this thread, except Android Auto so far. Case is, that connecting to MA needs VPN now, and Android Auto doesn't work while you're on VPN...

Yep, AA (Android Auto) would be beneficial. I do not have issues as of now with AA apps over VPN (ovpn) - Anyways, it took 'ages' to properly set the VPN server+profiles. :D

**Reply by formatBCE** [reply]
> > I actually implemented all the things from this thread, except Android Auto so far. Case is, that connecting to MA needs VPN now, and Android Auto doesn't work while you're on VPN...
> 
> Yep, AA (Android Auto) would be beneficial. I do not have issues as of now with AA apps over VPN (ovpn) - Anyways, it took 'ages' to properly set the VPN server+profiles. :D

I did it, it works great.

**Reply by returnofthem** [reply]
Hey all, I've been looking for a solution to work around MA not having an Android app - and specifically one which provides media controls - and even more specifically one which works with Android Auto - so hearing something is in the works is great!

MA is an amazing system and I'd like to contribute back if I can.

I can't offer any coding skills, but if you'd like me to alpha/beta test anything related to an Android app being developed then I'd be happy to. 

Edit: the AA + VPN case is an interesting one I could maybe help test as I'm already running a wireguard VPN. No problems with other AA apps with the VPN so far.

**Reply by formatBCE** [reply]
@returnofthem sure. Come to MA Discord, there's GPlay registration in progress. :)
https://discord.com/channels/753947050995089438/1343766783768789002

**Reply by returnofthem** [reply]
@formatBCE just tried following that link but it looks like I don't have access - I can join the main HA server but can't access whatever area that is.

**Reply by formatBCE** [reply]
> @formatBCE just tried following that link but it looks like I don't have access - I can join the main HA server but can't access whatever area that is.

It's Music Assistant server, not HA.
https://discord.gg/Kaq5wJ62

**Comment by JiDW** [comment]
I'm wondering how paid providers would work in that setup. Most providers offer "duo" or "family" subscription options but right now anyone can use my Music Assistant with a single paid account but I think it is OK as long as it stays in the house. I would not like to see some providers cutting integrations because of some potential abuse so I'm a bit worried.

**Reply by OzGav** [reply]
Isn’t this up to you to manage so that not “anyone can use my Music Assistant”? Only my family can access my HA instance…?

**Reply by JiDW** [reply]
I'm not worried about me directly. I'm concerned that if a Music Assistant phone app easily enables bypassing some provider limitation, their respective company may take action against MA compatibility.

I have no idea tho, I'm just trying to understand how things works in such case. I don't even know if you can stream from the same provider (let's say Spotify or Qobuz) on two different devices.

**Reply by OzGav** [reply]
Fair enough but I don’t think this is any different from someone using their Spotify app on one device and Qobuz on another. We take piracy and the circumvention of ads seriously and don’t do anything to enable that.

**Reply by JiDW** [reply]
Oh but I know you do, that's why I was wondering.
I probably wasn't clear in my use case:

My wife and I can both access HA and MA. We currently pay for a Duo subscription at Qobuz so we can each listen to music when on the road or at work. 
On MA, we have my own Qobuz account as a music provider that we use for playing musics in the house.

My initial question was wondering how that would work, should we be able to have a Phone app to use MA outside the house, that's all.

**Reply by OzGav** [reply]
You can already do that now without the phone app by using the web player.

**Reply by formatBCE** [reply]
Frankly, with some setup (like VPN) you already can do this on any player. The fact that MA allows me using same Spotify account at home (via MA) and in the car (via Spotify app) simultaneously - is basically same violation (?).

**Comment by swever826** [comment]
I was looking for a possible mobile app and found someone that did it already for Android! Looks cool if you guys want to check it out

https://github.com/CollotsSpot/Ensemble

**Reply by formatBCE** [reply]
I saw it, and even encouraged developer to help with official one. He refused.

**Reply by howels** [reply]
It also appears to be entirely coded with AI and the author isn't interested in working with F-Droid. Seeing the entire commit history authored by "Your Name and claude" doesn't make me confident.

**Reply by formatBCE** [reply]
Yeah exactly.
It's Flutter too.

**Reply by swever826** [reply]
Oh wow I didn't know all that :s I didn't even know there was an official app in the works haha I'll use the official one instead once it's fixed, since 2.7 version broke it.

Thanks for the info! I'll leave the link of the official app repo if someone stumbles upon this thread and didn't know about it either:
https://github.com/music-assistant/kmp-client-app

**Reply by feerlessleadr** [reply]
Do you happen to have a link to the discord for your app? The link above is expired. Thanks

**Reply by formatBCE** [reply]
> Do you happen to have a link to the discord for your app? The link above is expired. Thanks

https://discord.com/channels/753947050995089438/1343766783768789002

**Reply by feerlessleadr** [reply]
Thanks, for some reason that link doesn't work.

**Reply by formatBCE** [reply]
> Thanks, for some reason that link doesn't work. 

Are you on the Music Assistant Discord?

**Reply by formatBCE** [reply]
https://discord.gg/9b6g3bZde

**Reply by feerlessleadr** [reply]
Thank you! That did it!

**Comment by bfranske** [comment]
I'd also like to find a unified solution for playing at home (on players) and while out and about (on phones and android auto/carplay). For playback on phones I am currently using Navidrome with Subsonic phone clients because I do sometimes travel in areas with poor or no phone service so the ability to cache songs locally is important. It seems like the easiest path would be to just support Subsonic clients from Music Assistant but a better solution might be a dedicated phone app solution (or PWA I suppose) but I do want to clarify that for me it's important the local caching of some reasonable amount of music is important for when traveling in areas with limited/no service.

FWIW I want to unify my music playback solution so that playlists can be synchronized (i.e. I can build a playlist at home for use on a player but also play the same playlist on a mobile device).

**Reply by cociweb** [reply]
> I'd also like to find a unified solution for playing at home (on players) and while out and about (on phones and android auto/carplay). For playback on phones I am currently using Navidrome with Subsonic phone clients because I do sometimes travel in areas with poor or no phone service so the ability to cache songs locally is important. It seems like the easiest path would be to just support Subsonic clients from Music Assistant but a better solution might be a dedicated phone app solution (or PWA I suppose) but I do want to clarify that for me it's important the local caching of some reasonable amount of music is important for when traveling in areas with limited/no service.
> 
> FWIW I want to unify my music playback solution so that playlists can be synchronized (i.e. I can build a playlist at home for use on a player but also play the same playlist on a mobile device).

In my opinion, you can use for these purposes the web frontend. on your mobile you can create an app from chrome/edge (thanks to sendspin from 2.7 upwards.)

The only usecase what I see is the android auto. This is the most relevant and most missed feature.

**Comment by howels** [comment]
A phone app could provide caching for the album art and the full list of artists and albums currently the web app doesn't scroll properly - the scrollbar only corresponds to the portion of the library that has been loaded so far in the current web session. So scroll to the bottom and you are not actually at the end of the list, only just the end of what has loaded so far.
What is more, the web app often needs to reconnect and doesn't display the previous data while reconnecting. On android a web app does not appear to be a substitute for native apps in terms of integrations, such as offering persistent UI, or persistent notifications or using the physical volume keys to control the MA volume levels. Web apps cannot hijack the volume keys AFAIK and are liable to unload when running in the background and only reconnect when the page is a foreground process.
Maybe I just haven't seen a good example of a web app, if these can be addressed by a web app that would be great.

**Reply by formatBCE** [reply]
You wrote many good points. Working on bunch.
